File: src/main/java/com/flexpoker/signup/repository/InMemorySignUpRepository.java
package com.flexpoker.signup.repository;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Repository;

import com.flexpoker.signup.SignUpUser;

@Profile("default")
@Repository
public class InMemorySignUpRepository implements SignUpRepository {

    private final Set<String> usernameSet;

    private final Map<UUID, Map<String, UUID>> signUpCodeMap;

    private final Map<UUID, SignUpUser> signUpUserMap;

    public InMemorySignUpRepository() {
        usernameSet = new HashSet<>();
        signUpCodeMap = new HashMap<>();
        signUpUserMap = new HashMap<>();
        addDefaultSignUps();
    }

    @Override
    public boolean usernameExists(String username) {
        return usernameSet.contains(username);
    }

    @Override
    public boolean signUpCodeExists(UUID signUpCode) {
        return signUpCodeMap.containsKey(signUpCode);
    }

    @Override
    public UUID findAggregateIdByUsernameAndSignUpCode(String username, UUID signUpCode) {
        Map<String, UUID> foundSignUpCode = signUpCodeMap.get(signUpCode);
        if (foundSignUpCode != null) {
            UUID aggregateId = foundSignUpCode.get(username);
            return aggregateId;
        }

        return null;
    }

    @Override
    public void storeSignUpInformation(UUID aggregateId, String username, UUID signupCode) {
        signUpCodeMap.put(signupCode, new HashMap<>());
        signUpCodeMap.get(signupCode).put(username, aggregateId);
    }

    @Override
    public void storeNewlyConfirmedUsername(String username) {
        usernameSet.add(username);
    }

    @Override
    public UUID findSignUpCodeByUsername(String username) {
        return signUpCodeMap.entrySet().stream()
                .filter(x -> x.getValue().containsKey(username))
                .findAny().get().getKey();
    }

    private void addDefaultSignUps() {
        storeNewlyConfirmedUsername("player1");
        storeNewlyConfirmedUsername("player2");
        storeNewlyConfirmedUsername("player3");
        storeNewlyConfirmedUsername("player4");
    }

    @Override
    public SignUpUser fetchSignUpUser(UUID signUpUserId) {
        return signUpUserMap.get(signUpUserId);
    }

    @Override
    public void saveSignUpUser(SignUpUser signUpUser) {
        signUpUserMap.put(signUpUser.getAggregateId(), signUpUser);
    }

}

File: src/main/kotlin/com/flexpoker/signup/repository/InMemorySignUpRepository.kt
package com.flexpoker.signup.repository

import com.flexpoker.config.ProfileNames
import com.flexpoker.signup.SignUpUser
import com.flexpoker.util.encodePassword
import jakarta.annotation.PostConstruct
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Repository
import java.util.UUID

@Profile(ProfileNames.DEFAULT, ProfileNames.SIGNUP_INMEMORY)
@Repository
class InMemorySignUpRepository : SignUpRepository {

    private val usernameSet: MutableSet<String> = HashSet()
    private val signUpCodeMap: MutableMap<UUID, MutableMap<String, UUID>> = HashMap()
    private val signUpUserMap: MutableMap<UUID, SignUpUser> = HashMap()

    override fun usernameExists(username: String): Boolean {
        return usernameSet.contains(username)
    }

    override fun signUpCodeExists(signUpCode: UUID): Boolean {
        return signUpCodeMap.containsKey(signUpCode)
    }

    override fun findAggregateIdByUsernameAndSignUpCode(username: String, signUpCode: UUID): UUID? {
        val foundSignUpCode: Map<String, UUID>? = signUpCodeMap[signUpCode]
        return foundSignUpCode?.get(username)
    }

    override fun storeSignUpInformation(aggregateId: UUID, username: String, signUpCode: UUID) {
        signUpCodeMap[signUpCode] = HashMap()
        signUpCodeMap[signUpCode]!![username] = aggregateId
    }

    override fun storeNewlyConfirmedUsername(username: String) {
        usernameSet.add(username)
    }

    override fun findSignUpCodeByUsername(username: String): UUID {
        return signUpCodeMap.entries.stream()
            .filter { x: Map.Entry<UUID, Map<String, UUID>> -> x.value.containsKey(username) }
            .findAny().get().key
    }

    override fun fetchSignUpUser(signUpUserId: UUID): SignUpUser? {
        return signUpUserMap[signUpUserId]
    }

    override fun saveSignUpUser(signUpUser: SignUpUser) {
        signUpUserMap[signUpUser.aggregateId] = signUpUser
    }

    @PostConstruct
    private fun addDefaultSignUps() {
        addDefaultSignUp("player1")
        addDefaultSignUp("player2")
        addDefaultSignUp("player3")
        addDefaultSignUp("player4")
    }

    private fun addDefaultSignUp(username: String) {
        val id = UUID.randomUUID()
        val signUpCode = UUID.randomUUID()
        storeNewlyConfirmedUsername(username)
        saveSignUpUser(SignUpUser(id, signUpCode, "$username@test.com", username, encodePassword(username)))
        storeSignUpInformation(id, username, signUpCode)
    }

}